blog: Add Supabase libraries LWX article (#19761)
supabase/supabase: apps/www/_blog/2023-12-15-v2-client-libraries.mdx

---
title: 'Supabase Libraries V2: Python, Swift, Kotlin, Flutter, and Typescript'

> description: 'Swift, Kotlin, C#, and Python are now stable and are onto the same version number v2.'

Reviewer: ```suggestion
description: 'Swift, Kotlin, C#, and Python are now stable and moving to the v2 API.'
```

launchweek: X
categories:
  - product
tags:
  - launch-week
date: '2023-12-15'
published_at: '2023-12-15:00:05.000-07:00'
toc_depth: 3
author: paul_copplestone
image: lwx-client-libs/client-libs-OG.png
thumb: lwx-client-libs/client-libs-thumb.png
---

Here is what you need to know:

> - [Swift](https://supabase.com/docs/reference/swift/introduction), [Kotlin](https://supabase.com/docs/reference/kotlin/introduction), [C#](https://supabase.com/docs/reference/csharp/introduction), and [Python](https://supabase.com/docs/reference/python/introduction) are now stable.

Reviewer: ```suggestion
- [Swift](/docs/reference/swift/introduction), [Kotlin](/docs/reference/kotlin/introduction), [C#](/docs/reference/csharp/introduction), and [Python](/docs/reference/python/introduction) are now stable.
```

> - We’ve added even more support for mobile using React Native and [Expo](https://supabase.com/docs/guides/getting-started/tutorials/with-expo-react-native).

Reviewer: ```suggestion
- We’ve added even more support for mobile using React Native and [Expo](/docs/guides/getting-started/tutorials/with-expo-react-native).
```

- We’re consolidating all client libraries onto the same version number (v2) so that you get a consistent API across all libraries.

## Version 2, for all libraries

> All of the libraries mentioned in this post are now on v2. This is because we want the API for each library to “move in lock step”. You should be able to jump around each of the client libraries and they should operate the same, barring any language idiosyncrasies.

Reviewer: ```suggestion
All of the libraries mentioned in this post are now on v2. We want the API for each library to “move in lock step”. You should be able to jump around each of the client libraries and they should operate the same, barring any language idiosyncrasies.
```

## Supabase Python v2

Supabase Python is now stable thanks to the following maintainers: [Anand](https://github.com/anand2312), [Daniel Reinón García](https://github.com/dreinon), [Leynier Gutiérrez González](https://github.com/leynier), [Joel Lee](https://github.com/J0), and [Andrew Smith](https://github.com/silentworks/).

Check out the [docs](https://supabase.com/docs/reference/python/introduction), as well as these Python examples to help you get started:

- [Slack Consolidate: a slackbot built with Python and Supabase](https://supabase.com/blog/slack-consolidate-slackbot-to-consolidate-messages)
- [Python data loading with Supabase](https://supabase.com/blog/loading-data-supabase-python)
- [GitHub OAuth in your Python Flask app](https://supabase.com/blog/oauth2-login-python-flask-apps)

<Admonition type="note" label="Vecs">

Don’t forget that we also have Vecs, a Python client for managing and querying vector stores in Postgres with the pgvector extension. [Get started](https://supabase.com/docs/guides/ai/python/api)!

> </Admonition>

Reviewer: For some reason this doesn't look like it does in wednesday's article?

https://supabase.com/blog/supabase-branching#data-seeding

Reviewer: <img width="728" alt="image" src="https://github.com/supabase/supabase/assets/10214025/cd01f59d-8ff3-4bac-8392-913da39fcb06">
<img width="803" alt="image" src="https://github.com/supabase/supabase/assets/10214025/51e50a80-190a-470e-901e-42957355c95c">

  - [Android Native Google login & in-app OAuth (Android)](https://github.com/supabase-community/supabase-kt/tree/master/demos/android-login)
  - [Multi-Factor Authentication (Desktop/Android/Browser)](https://github.com/supabase-community/supabase-kt/tree/master/demos/multi-factor-authentication)
  - [Multiplatform Deep Linking (Desktop/Android)](https://github.com/supabase-community/supabase-kt/tree/master/demos/multiplatform-deeplinks)
  - [Groceries Store App (Android)](https://github.com/hieuwu/android-groceries-store)

## Typescript v2 updates

We’ve made several updates for Typescript support in `supabase-js`:

- the Supabase CLI now generates [Table and Enum shorthands](https://supabase.com/docs/guides/api/rest/generating-types#type-shorthands)
- new helpers `QueryResult` and `QueryData` extract the result types for [complex queries and joins](https://supabase.com/docs/guides/api/rest/generating-types#response-types-for-complex-queries)
- relationships between tables are now supported:
  - one-to-many relationships are typed as `T[]`
  - many-to-one relationships are typed as `T | null`

## Flutter v2 updates

> The core theme of Flutter v2 has been stability and better DX. We want to give a shout-out to [Vinzent](https://github.com/Vinzent03), a community maintainer who has done the majority of the work. Some notable improvements:

Reviewer: ```suggestion
The core theme of Flutter v2 has been stability and better DX. Shout-out to [Vinzent](https://github.com/Vinzent03), a community maintainer who has done the majority of the work. Some notable improvements:
```

  - Lighter package size. We have removed some dependencies

## React Native and Expo support

We’ve [worked with the Expo team](https://docs.expo.dev/guides/using-supabase/) to improve support for React Native.

<div className="video-container">
  <iframe
    className="w-full"
    src="https://www.youtube-nocookie.com/embed/7DYZdAPPJOw"
    title="YouTube video player"
    allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
    allowfullscreen
  />
</div>

By default, `supabase-js` uses the browser's `localStorage` mechanism to persist the user's session. This can be extended with platform-specific implementations. React Native can target native mobile and web applications with the same code base, so we need a storage implementation that works for all these platforms. Now you can use [react-native-async-storage](https://supabase.com/blog/react-native-authentication#authentication-storage) or a combination with [expo-secure-sorage](https://supabase.com/blog/react-native-authentication#encrypting-the-user-session) for AES encrypted sessions.

> Beyond that we’ve focused on making supabase-js highly compatible with React Native and created plenty of [video tutorials](https://youtube.com/playlist?list=PL5S4mPUpp4OsrbRTx21k34aACOgpqQGlx&si=Ez-0S4QhBxtayYsq) and documentation for [crafting an offline-first experience](https://supabase.com/blog/react-native-offline-first-watermelon-db), [uploading files to Supabase Storage](https://supabase.com/blog/react-native-storage), and [sending push notifications](https://supabase.com/docs/guides/functions/examples/push-notifications).

Reviewer: ```suggestion
Beyond that we’ve focused on making supabase-js highly compatible with React Native and created plenty of [video tutorials](https://youtube.com/playlist?list=PL5S4mPUpp4OsrbRTx21k34aACOgpqQGlx&si=Ez-0S4QhBxtayYsq) and documentation for: 

- [crafting an offline-first experience](https://supabase.com/blog/react-native-offline-first-watermelon-db)
- [uploading files to Supabase Storage](https://supabase.com/blog/react-native-storage)
- [sending push notifications](https://supabase.com/docs/guides/functions/examples/push-notifications).
```

## Our approach to client libraries

We have a strong preference to develop the client libraries with our community. This is part of our open source philosophy:

### The Cathedral or the Bazaar?

If you haven’t already, it’s worth reading “[The Cathedral and the Bazaar](https://en.wikipedia.org/wiki/The_Cathedral_and_the_Bazaar)” by Eric S. Raymond. In short, it contrasts two software development approaches:

> 1. **The "Cathedral"** represents closed, centralized development, where a small group of developers work in isolation.
> 2. **The "Bazaar"** symbolizes open, decentralized collaboration, where a large community of developers openly shares and collaborates on code.

Reviewer: ```suggestion
1. The **Cathedral** represents closed, centralized development, where a small group of developers work in isolation.
2. The **Bazaar** symbolizes open, decentralized collaboration, where a large community of developers openly shares and collaborates on code.
```

We believe the “bazaar” model is the right model for an open source business. If you aren’t constantly pushing in this direction then it’s extremely likely that the company will relicense (seen most recently with [Hashi](https://news.ycombinator.com/item?id=37081306)). The way we see it, the more we can foster our community the less power we have.

### Fostering the community

> We just reached 1000 contributors to our [main repo](https://github.com/supabase/supabase). It’s not easy to build a community of contributors, it’s something that need to be fostered.

Reviewer: ```suggestion
We just reached 1000 contributors to our [main repo](https://github.com/supabase/supabase). It’s not easy to build a community of contributors, it’s something that need to be fostered. (Shout out to [@tobiastornros](https://github.com/tobiastornros), contributor #1000 - and everyone else who [contributed yesterday](https://x.com/kiwicopple/status/1735282666872987673?s=20))
```

> The client libs are one of the best ways to foster the community because they lower complexity than some of the tools we maintain (do you know [Haskell](https://github.com/PostgREST/postgrest)?).

Reviewer: ```suggestion
The client libs are one of the best ways to foster the community because they are lower-complexity than some of the tools we maintain (do you know [Haskell](https://github.com/PostgREST/postgrest)?).
```

<Img
  src={{
    dark: '/images/blog/lwx-client-libs/community-dark.png',
    light: '/images/blog/lwx-client-libs/community-light.png',
  }}
  alt="Diagram explaining Supabase community"
/>

> We want the Supabase community to outlive us. With more community maintainers, you should feel _even safer_ knowing that there is already a continuity plan in place. We’re [giving back to the community](https://supabase.com/blog/supabase-series-b#giving-back) as those contributions grow, and we hope to expand this as we become even more commercially successful.

Reviewer: ```suggestion
We want the Supabase community to outlive us. With more community maintainers, you should feel _even safer_ knowing that there is already a continuity plan in place. We’re [giving back to the community](https://supabase.com/blog/supabase-series-b#giving-back) and we hope to expand this as we become even more commercially successful.
```

### Modularity

> As a [Principle](https://supabase.com/docs/guides/getting-started/architecture#product-principles), we develop a single library for each tool we support. While Supabase may “feel” like a single tool when you’re using it, it is a set of individual tools which you can use independently (especially useful for self hosting):

Reviewer: ```suggestion
As a [Principle](https://supabase.com/docs/guides/getting-started/architecture#product-principles), we develop a library for each tool we support. While Supabase may “feel” like a single tool when you’re using it, it is a set of individual tools which you can use independently (especially useful for self hosting):
```
